src/ngraph/pass/algebraic_simplification.cpp

```diff
@@ -345,6 +347,86 @@ static bool is_input_uniform_constant(shared_ptr<Node> op,
     return rc;
 }
 
+//`simplify_gather`, optimizes gather if Gather is gathering the
+// whole input tensor
+static bool simplify_gather(std::shared_ptr<Node> node)
+{
+    auto gather = as_type_ptr<op::v1::Gather>(node);
+
+    if (gather)
+    {
+        // check if we are gathering the whole input
+        auto data = node->get_argument(0);
+        auto indices = node->get_argument(1);
+
+        // we need to know data and indices shape to infer if gather is Nop
+        if (data->get_output_partial_shape(0).is_dynamic() ||
```

Comment:
Algorithm should be the following:
1. get_axis() != AXIS_NOT_SET_VALUE && axis > 0
2. data_partial_shape[axis] is static
3. indices are 1D constant and contain range of values 0...len(data_partial_shape[axis]). case with single 1D index doesn't require it to be constant.

I'm I missing smth or we do not need such restrictions?

Reply:
1. when I query `gather->get_axis()`, the member function would do the check for point 1 and point 2 as you explained and infer the axis value. if its dynamic the axis value will be set to `AXIS_NOT_SET_VALUE`. 

2. For point 3, I am doing this check here, 
https://github.com/NervanaSystems/ngraph/blob/8edab564148392b14200ac0e906e2b7b46ec8121/src/ngraph/pass/algebraic_simplification.cpp#L398-#L405
if the indices are 1D and whose value values equal to {0..data_partial_shape[axis]}. then that would fetch the entire input tensor

Please let me know if i resolved your questions :)